src/Cornelis/Highlighting.hs

```diff
@@ -74,82 +76,73 @@ highlightBuffer b hs = do
+              $ go (incIndex pos') (incIndex line) ss
       | otherwise = mempty
 
-lookupPoint :: LineIntervals -> BufferOffset -> Maybe (LineNumber, Int64)
-lookupPoint (LineIntervals im) off = do
-  (IM.Interval scs _, (startline, s)) <- listToMaybe $ IM.search off im
-  pure ( startline
-       , fromIntegral $ toBytes s (offsetDiff off scs) - 1
-       )
-
+lookupPoint :: LineIntervals -> AgdaIndex -> Maybe VimPos
+lookupPoint (LineIntervals im) i = do
+  (IM.Interval lineStart _, (line, s)) <- listToMaybe $ IM.search i im
+  let col = toBytes s (zeroIndexed 0 .+ (i .-. lineStart))
+  pure (Pos line col)
 
 ------------------------------------------------------------------------------
 -- | Returns any holes it tried to highlight on the left
 addHighlight
     :: Buffer
     -> LineIntervals
     -> Highlight
-    -> Neovim CornelisEnv ([Interval' Int64], Maybe Extmark)
+    -> Neovim CornelisEnv ([VimInterval], Maybe Extmark)
 addHighlight b lis hl = do
-  case (,)
-          <$> lookupPoint lis (hl_start hl)
-               -- Subtract 1 here from the end offset because unlike everywhere
-               -- else in vim, extmark ranges are inclusive...
-           <*> lookupPoint lis (offsetDiff (hl_end hl) (Offset 1)) of
-    Just (start@(sl, sc), end@(el, ec)) -> do
+  debug hl
+  case Interval
+           <$> lookupPoint lis (hl_start hl)
+           <*> lookupPoint lis (hl_end hl) of
+    Just (int@(Interval start end@(Pos el ec))) -> do
       let atom = fromMaybe "" $ listToMaybe $ hl_atoms hl
-      ext <- setHighlight b (first (fromIntegral . getOneIndexedLineNumber) start)
-                            (first (fromIntegral . getOneIndexedLineNumber) end)
-                          $ hlGroup atom
+      ext <- setHighlight b int $ hlGroup atom
 
       pure $ (, ext) $ case atom == "hole" of
         False -> []
         True ->
-          pure $ Interval
-            { iStart = Pos (incLineNumber sl) sc
-            , iEnd   = Pos (incLineNumber el) (ec + 1)
-            }
+          pure $ Interval start (Pos el (incIndex ec))
     Nothing -> pure ([], Nothing)
 
 
 setHighlight
     :: Buffer
-    -> (Int64, Int64)
-    -> (Int64, Int64)
+    -> Interval VimPos
     -> HighlightGroup
     -> Neovim CornelisEnv (Maybe Extmark)
-setHighlight b (sl, sc) (el, ec) hl = do
+setHighlight b x@(Interval (Pos sl sc) (Pos el ec)) hl = do
   ns <- asks ce_namespace
-  flip catchNeovimException (const (pure Nothing))
-    $ fmap (Just . coerce) $ nvim_buf_set_extmark b ns sl sc $ M.fromList
+  let zi = fromZeroIndexed
+  y <- flip catchNeovimException (const (pure Nothing))
+    $ fmap (Just . coerce) $ nvim_buf_set_extmark b ns (zi sl) (zi sc) $ M.fromList
     [ ( "end_line"
-      , ObjectInt el
+      , ObjectInt (zi el)
       )
       -- unlike literally everywhere else in vim, this function is INCLUSIVE
       -- in its end column
     , ( "end_col"
-      , ObjectInt $ ec + 1
+      , ObjectInt $ zi ec
```

Comment:
Where did the +1 go?

Reply:
The `+1` cancels out with `(offsetDiff (hl_end hl) (Offset 1)) ` at the old line 119